File: js/src/strategy.ts
/** Candidate ordering strategies. A candidate is one (provider, key, model). */
import type { KeyState } from "./keys.js";

export const STRATEGIES = ["priority", "round_robin", "quota_aware", "latency"] as const;
export type Strategy = (typeof STRATEGIES)[number];

export interface Candidate {
  provider: any;
  key: KeyState;
  model: string;
}

export function orderCandidates(
  providers: any[],
  alias: string | string[],
  now: number,
  strategy: string,
  rr: { p: number },
): Candidate[] {
  let provs = [...providers];

  // provider `priority` is the universal tiebreak: primary for priority,
  // secondary for the dynamic strategies, baseline order for round_robin.
  if (strategy === "round_robin" && provs.length) {
    provs.sort((a, b) => a.priority - b.priority);
    const i = (rr.p ?? 0) % provs.length;
    provs = [...provs.slice(i), ...provs.slice(0, i)];
    rr.p = (rr.p ?? 0) + 1;
  } else if (strategy === "quota_aware") {
    provs.sort((a, b) => b.capacity(now) - a.capacity(now) || a.priority - b.priority);
  } else if (strategy === "latency") {
    // Infinity - Infinity is NaN (falsy) -> the priority tiebreak kicks in
    provs.sort((a, b) => a.avgLatency() - b.avgLatency() || a.priority - b.priority);
  } else {
    provs.sort((a, b) => a.priority - b.priority);
  }

  // Build each provider's own ordered sublist (rotated keys, then models).
  const perProvider: Candidate[][] = [];
  for (const p of provs) {
    let keys = [...p.keys];
    if (keys.length) {
      const ki = p._rr % keys.length;
      keys = [...keys.slice(ki), ...keys.slice(0, ki)];
      p._rr++;
    }
    const models = p.resolveModels(alias);
    const sub: Candidate[] = [];
    for (const k of keys) for (const mid of models) sub.push({ provider: p, key: k, model: mid });
    if (sub.length) perProvider.push(sub);
  }

  // Interleave breadth-first across providers (best model of each, then next).
  const out: Candidate[] = [];
  const maxLen = perProvider.reduce((m, s) => Math.max(m, s.length), 0);
  for (let rank = 0; rank < maxLen; rank++) {
    for (const sub of perProvider) if (rank < sub.length) out.push(sub[rank]);
  }
  return out;
}

File: src/freelm/strategy.py
"""Candidate ordering strategies.

A *candidate* is one concrete (provider, key, model) we could try. Strategies
decide the order; the engine walks them and picks the first that is ready.
"""
from __future__ import annotations

from dataclasses import dataclass
from itertools import zip_longest
from typing import Any, Dict, List

PRIORITY = "priority"
ROUND_ROBIN = "round_robin"
QUOTA_AWARE = "quota_aware"
LATENCY = "latency"
STRATEGIES = (PRIORITY, ROUND_ROBIN, QUOTA_AWARE, LATENCY)


@dataclass
class Candidate:
    provider: Any  # providers.base.Provider
    key: Any       # _keys.KeyState
    model: str


def order_candidates(
    providers: List[Any],
    alias: str,
    now: float,
    strategy: str,
    rr: Dict[str, int],
) -> List[Candidate]:
    """Ordered list of (provider, key, model) candidates to try.

    Providers are ranked by ``strategy``; candidates are then **interleaved
    breadth-first across providers** — i.e. the best model of every provider is
    tried before any provider's second model. This guarantees failover reaches
    every provider quickly instead of burning all attempts on one provider's
    many (possibly throttled) models.
    """
    provs = list(providers)

    if strategy == ROUND_ROBIN and provs:
        i = rr.get("p", 0) % len(provs)
        provs = provs[i:] + provs[:i]
        rr["p"] = rr.get("p", 0) + 1
    elif strategy == QUOTA_AWARE:
        provs.sort(key=lambda p: p.capacity(now), reverse=True)
    elif strategy == LATENCY:
        provs.sort(key=lambda p: p.avg_latency())
    else:  # PRIORITY
        provs.sort(key=lambda p: p.priority)

    # Build each provider's own ordered sublist (rotated keys, then models).
    per_provider: List[List[Candidate]] = []
    for p in provs:
        keys = list(p.keys)
        if keys:  # rotate keys within a provider to spread load
            ki = p._rr % len(keys)
            keys = keys[ki:] + keys[:ki]
            p._rr += 1
        models = p.resolve_models(alias)
        sub = [Candidate(p, k, mid) for k in keys for mid in models]
        if sub:
            per_provider.append(sub)

    # Interleave: rank-0 of every provider first (in strategy order), then rank-1, ...
    candidates: List[Candidate] = []
    for rank in zip_longest(*per_provider):
        for c in rank:
            if c is not None:
                candidates.append(c)
    return candidates
